source/common/unified_memory_allocation/src/memory_pool.c

```diff
@@ -6,43 +6,101 @@
-    struct uma_memory_pool_ops_t ops;
     void *pool_priv;
+
+    struct uma_memory_pool_ops_t ops;
+
+    struct uma_memory_provider_desc_t *providers;
+    size_t numProviders;
 };
 
-enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, void *params,
-                                uma_memory_pool_handle_t *hPool) {
+enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, struct uma_memory_provider_desc_t *providers,
+                                size_t numProviders, void *params, uma_memory_pool_handle_t *hPool) {
+    if (providers && !numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!providers && numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    for (size_t i = 0; i < numProviders; i++) {
+        // UMA_MEMORY_PROVIDER_TYPE_TRACKER can only be created by the UMA library
+        if (providers[i].providerType == UMA_MEMORY_PROVIDER_TYPE_TRACKER) {
+            return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    enum uma_result_t ret = UMA_RESULT_SUCCESS;
     uma_memory_pool_handle_t pool = malloc(sizeof(struct uma_memory_pool_t));
     if (!pool) {
         return UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
     }
 
     assert(ops->version == UMA_VERSION_CURRENT);
 
-    pool->ops = *ops;
+    pool->providers = calloc(numProviders ? numProviders : 1, sizeof(struct uma_memory_provider_desc_t));
+    if (!pool->providers) {
+        ret = UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_providers_alloc;
+    }
 
-    void *pool_priv;
-    enum uma_result_t ret = ops->initialize(params, &pool_priv);
-    if (ret != UMA_RESULT_SUCCESS) {
-        free(pool);
-        return ret;
+    size_t providerInd = 0;
+    if (!numProviders) {
+        pool->numProviders = 1;
+        pool->providers[0].providerType = UMA_MEMORY_PROVIDER_TYPE_TRACKER;
+        ret = umaTrackingMemoryProviderCreate(NULL, pool, &pool->providers[0].hProvider);
+        if (ret != UMA_RESULT_SUCCESS) {
+            goto err_providers_init;
+        }
+    } else {
+        pool->numProviders = numProviders;
+        for (providerInd = 0; providerInd < numProviders; providerInd++) {
+            pool->providers[providerInd].providerType = providers[providerInd].providerType;
+            ret = umaTrackingMemoryProviderCreate(providers[providerInd].hProvider, pool, &pool->providers[providerInd].hProvider);
+            if (ret != UMA_RESULT_SUCCESS) {
+                goto err_providers_init;
+            }
+        }
     }
 
-    pool->pool_priv = pool_priv;
+    pool->ops = *ops;
+    ret = ops->initialize(pool->providers, pool->numProviders, params, &pool->pool_priv);
+    if (ret != UMA_RESULT_SUCCESS) {
+        goto err_pool_init;
+    }
 
     *hPool = pool;
-
     return UMA_RESULT_SUCCESS;
+
+err_pool_init:
+err_providers_init:
+    for (size_t i = 0; i < providerInd; i++) {
+        umaMemoryProviderDestroy(pool->providers[i].hProvider);
```

Comment:
I think we should not destroy memory providers that were not created by us.
If it is created and passed by the user as a parameters to `umaPoolCreate` function than user is responsible to destroy them.

Reply:
Yes, and we are not destroying the user-provider providers - just the wrappers (trackingMemoryProviders) that were created by UMA.

Reply:
Ahh, got it. Could you please add a comment?

Reply:
Done. I've also added description to umaTrackingMemoryProviderCreate